GVFS/GVFS.Virtualization/FileSystemCallbacks.cs

```diff
@@ -180,7 +173,11 @@ public bool TryStart(out string error)
 
             this.backgroundFileSystemTaskRunner.Start();

```

Comment:
I noticed that we don't have any functional tests for loading persisted background tests (and I've filed #816 to add some).  

Could you try running a few manual tests to validate that this still works as expected with this change? (I took a look through the background task code and it all looked like it should be OK to me).

Reply:
I was looking at this as well and the 2 background tasks that are doing file operations are the folder rename and the folder predelete.  If the file virtualization is not up when those run what will be the results?  

Looks like for the folder rename it would only not be removing entries from the modified paths which wouldn't be a big deal also the newly created list would be empty at this point anyways so I think we are okay there.

For the predelete if it didn't find the folder because the virtualization wasn't up yet and it removed it from the modified path would be an issue but could this not happen because if we are getting a predelete the placeholder is on disk and it would find it?

Reply:
Great questions on those, please double check my thoughts on them for me 😄 

*Folder rename*

- We'll be enumerating a full folder (on Windows, folder rename will work that way once it's fixed on Mac, see #515) which should work fine if the virtualization root has not been started.  
- ModifiedPath.dat itself is outside the root so modifying it should be OK.

*Folder PreDelete*

- VFS4G will only check if the directory exists if the path is in `newlyCreatedFileAndFolderPaths`, and because VFS4G is just starting up `newlyCreatedFileAndFolderPaths` will be empty.
- Well, I guess that's not 100% guaranteed because another persisted task could have added the path(s) to `newlyCreatedFileAndFolderPaths`.
- However, any path that has been added to `newlyCreatedFileAndFolderPaths` should actually be on disk or not (i.e. it should not be pure virtual) and I would expect checking for existence to work properly even if the virtualization root is not started.  It would be worth confirming the behavior of `Directory.Exists` on Mac and Windows when the repo is not mounted.
- ModifiedPath.dat itself is outside the root so modifying it should be OK.

So for both of these scenarios I think we're okay now, but I could see accidentally breaking them if the background processing changes.  It might be worth adding a few comments near these background tasks as a reminder that they shouldn't assume the repo is mounted.

The other general thought I had was that ProjFS should not be triggering callbacks for I/O VFS4G initiates itself anyway, and so as long as the BG tasks aren't doing anything that ProjFS would block if we're not mounted I think we should be okay.

Reply:
@wilbaker @kewillford  please let me know if I'm testing this properly.

I put a breakpoint so we halt before` if (!this.fileSystemVirtualizer.TryStart(this, out error))`

move UnitTests UnitTests2
The provider that supports file system virtualization is temporarily unavailable.

If I Delete the UnitTests folder however, it appears again when I continue the mount.  I.e. the delete isn't remembered.

Reply:
@jeschu1 it sounds like you're trying to test a different scenario than I had in mind.  

To test loading BG tasks on mount the best approach I know is:

1. Temporarily modify the VFS4G source code so that it does not process BG tasks (e.g. make `BackgroundFileSystemTaskRunner.RunCallbackUntilSuccess` an infinite loop).

2. Perform the actions you'd like to have run on startup:
- Create a new directory and rename it
- Delete a directory (note that we only receive PreDelete for directories on Mac)

3. Stop GVFS.Mount.exe

4. There should now have a .gvfs\databases\BackgroundGitOperations.dat file on disk that has all of the unprocessed BG tasks.

5. Back out the change in 1 and mount the repo (it should process the BG tasks in the file)

The approach above could be used to add a functional test that validates that VFS4G can run the persisted BG tasks.  `BackgroundGitOperations.dat` could be checked into the repo and then moved into fresh VFS4G clones during the functional tests.

Reply:
@wilbaker new background task test added as described